Library/Homebrew/test/caveats_spec.rb

```diff
@@ -26,4 +27,123 @@ def caveats
+        end
+        expect(described_class.new(f).caveats).to include("startup:\n  sudo brew")
+      end
+
+      it "prints plist login information when f.plist_startup is nil" do
+        f = formula do
+          url "foo-1.0"
+          def plist
+            "plist_test.plist"
+          end
+        end
+        expect(described_class.new(f).caveats).to include("login:\n  brew")
+      end
+    end
+
+    context "when f.keg_only is not nil" do
+      Path = Pathname.new("path")
+
+      ## Refactoring needed:
+      # The following lines of code have been repeated in every example:
+      #   f = formula do
+      #       url "foo-1.0"
+      #       keg_only "some reason"
+      #   end
+      #   caveats = described_class.new(@f).caveats
+
+      it "tells formula is keg_only" do
+        f = formula do
+          url "foo-1.0"
+          keg_only "some reason"
+        end
+        caveats = described_class.new(f).caveats
+        expect(caveats).to include("keg-only")
+      end
+
+      it "gives command to be run when f.bin is a directory" do
+        f = formula do
+          url "foo-1.0"
+          keg_only "some reason"
+        end
+
+        allow(f).to receive(:bin).and_return(Path)
+        allow(f.bin).to receive(:directory?).and_return(true)
+
+        caveats = described_class.new(f).caveats
+
+        expect(caveats).to include(f.opt_bin.to_s)
+      end
+
+      it "gives command to be run when f.sbin is a directory" do
+        f = formula do
+          url "foo-1.0"
+          keg_only "some reason"
+        end
+
+        allow(f).to receive(:sbin).and_return(Path)
+        allow(f.sbin).to receive(:directory?).and_return(true)
+
+        caveats = described_class.new(f).caveats
+
+        expect(caveats).to include(f.opt_sbin.to_s)
+      end
+
+      context "when f.lib or f.include is a directory" do
+        it "gives command to be run when f.lib is a directory" do
+          f = formula do
+            url "foo-1.0"
+            keg_only "some reason"
+          end
+
+          allow(f).to receive(:lib).and_return(Path)
+          allow(f.lib).to receive(:directory?).and_return(true)
+
+          caveats = described_class.new(f).caveats
+
+          expect(caveats).to include("-L#{f.opt_lib}\n")
+        end
+
+        it "gives command to be run when f.include is a directory" do
+          f = formula do
+            url "foo-1.0"
+            keg_only "some reason"
+          end
+
+          allow(f).to receive(:include).and_return(Path)
+          allow(f.include).to receive(:directory?).and_return(true)
+
+          caveats = described_class.new(f).caveats
+
+          expect(caveats).to include("-I#{f.opt_include}\n")
+        end
+
+        it "gives PKG_CONFIG_PATH when f.share/\"pkgconfig\" is a directory" do
+          f = formula do
+            url "foo-1.0"
+            keg_only "some reason"
+          end
+
+          Pathname.new(f.share/"pkgconfig").mkpath
```

Comment:
I'd consider using `mkpath` in all the above examples, too, instead of stubbing.